sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc

```diff
@@ -0,0 +1,378 @@
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 7 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+The complex marray extension builds on the `sycl::ext::oneapi::complex` class,
+therefore this extension is dependent on
+link:sycl_ext_oneapi_complex.asciidoc[sycl_ext_oneapi_complex].
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+[NOTE]
+====
+This extension is not currently implemented in {dpcpp}.
+====
+
+
+== Overview
+
+Following the proposal for `sycl::ext::oneapi::complex`, which adds the support
+of complex for SYCL, this proposal allows for adding new complex math features.
+
+This document proposes the specialization of `sycl::marray` to add support for
+storing complex numbers in arrays.
+
+The proposal includes overloading the existing math functions to support complex
+marrays and adding member functions to simplify accessing, setting marray
+values, and constructing complex marrays.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_COMPLEX_MARRAY` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== Marray Complex Class Specialization
+
+The user interface of the `sycl::marray<sycl::ext::oneapi::complex, N>` class is
+changed compared to the SYCL-2020 generic `sycl::marray` interface.
+
+This proposition changes the marray interface when specialized for the
+`sycl::ext::oneapi::complex` type.
+
+The marray complex specialization is trivially copyable, and the type trait
+`is_device_copyable` should resolve to `std::true_type`.
+
+The marray definition used within this proposal assumes that any operator the
+`sycl::marray` class defines is only implemented if the marray's value type also
+implements the operator.
+
+For example, `sycl::marray<sycl::ext::oneapi::complex, N>` does not implement
+the modulus operator as `sycl::ext::oneapi::complex` does not support it.
+
+```C++
+namespace sycl {
+namespace ext {
+namespace oneapi {
```

Comment:
The [complex extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_complex.asciidoc) is experimental, so this should be in the experimental namespace too as it relies on it.

Reply:
The templated class `marray` has been declared and defined under the namespace `sycl/ext/oneapi`.

Correct me if I'm wrong, but an explicit specialization can only be done in the same namespace scope as the primary template.

Reply:
Ah, I see what you mean. I suppose this falls into the category of extension behavior of existing interfaces. Tag @gmlueck & @rolandschulz .